### PR TITLE
Remove use of `pika::for_loop` from permutations

### PR DESCRIPTION
**Pull request by msimberg**
Replaces uses of `pika::for_loop` with `pika::execution::experimental::bulk`. Since I've done this somewhat "blindly" (i.e. without fully understanding what the algorithm itself does) I'm not sure if the names I've used are good.

This doesn't seem to affect performance at all, presumably because this is only a tiny fraction of the tridiagonal eigensolver runtime to start with.

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
The stdexec build needs fixing, but this should otherwise be ready for reviews.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
In terms of functionality this requires pika 0.17.0 but because of various bugs in the sender functionalities this requires pika main. I've currently set CI to use `pika@main` to test that it works and it does. This PR can wait until 0.18.0 before being merged since it doesn't affect performance.

File: spack/packages/dla-future/package.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ class DlaFuture(CMakePackage, CudaPackage, ROCmPackage):
 
     depends_on("pika@0.15.1:", when="@0.1")
     depends_on("pika@0.16:", when="@0.2.0:")
+    depends_on("pika@main", when="@main")
```
Comment (msimberg) [inline spack/packages/dla-future/package.py:59]:
```suggestion
    depends_on("pika@0.18:", when="@main")
```

File: ci/.gitlab-ci.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ stages:
   timeout: 6 hours
   variables:
     GIT_SUBMODULE_STRATEGY: recursive
-    SPACK_SHA: b313b28e64c15761be0d45a16c922c25b2786f76
+    SPACK_SHA: 29aa7117f42f758bc537e03e4bedf66ced0accfa
```
Comment (msimberg) [inline ci/.gitlab-ci.yml:20]: Update to include pika 0.18.0.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by rasolca** [comment]
Some conflicts to be fixed

**Comment by rasolca** [comment]
Question: are we `pika_algorithm` free after we merge this PR?
If yes, please remove it from the dependencies.

File: spack/packages/dla-future/package.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ class DlaFuture(CMakePackage, CudaPackage, ROCmPackage):
 
     depends_on("pika@0.15.1:", when="@0.1")
     depends_on("pika@0.16:", when="@0.2.0:")
+    depends_on("pika@0.18:", when="@main")
```
[the remark this answers is not in the document]
Comment (msimberg) [reply spack/packages/dla-future/package.py]: Grr, thanks. Fixed: https://github.com/eth-cscs/DLA-Future/pull/945/commits/131ba91d755acad07726c8266fbb53cc93b22f73.

**Comment by msimberg** [comment]
> Question: are we `pika_algorithm` free after we merge this PR? If yes, please remove it from the dependencies.

We are now that #977 was merged :) Updated in https://github.com/eth-cscs/DLA-Future/pull/945/commits/4e5938346a4925928945e371aa1658b85691ed98.

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]

**Review by albestro — APPROVED** [review]
I'm not super happy that we now have a custom permutation for each case, but I don't have alternatives at the moment. So I'm fine with this 😄

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
This will need a rerun of the tests once daint is back in action. The test jobs are currently failing to start.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by rasolca** [comment]
I stopped the CI as this PR will require fixing conflicts after #964 gets merged.

**Comment by rasolca** [comment]
You can now fix the conflicts and run tests.

**Comment by msimberg** [comment]
cscs-ci run

**Pull request merged**